src/compiler/checker.ts

```diff
@@ -12893,6 +12893,91 @@ namespace ts {
+                        case SyntaxKind.NumericLiteral:
+                        case SyntaxKind.Identifier:
+                            memberName = (member.name as LiteralExpression | Identifier).text;
+                            break;
+                        default:
+                            continue;
+                    }
+                }
+
+                const static = forEach(member.modifiers, m => m.kind === SyntaxKind.StaticKeyword);
+                const names = static ? staticNames : instanceNames;
+                switch (member.kind) {
+                    case SyntaxKind.Constructor:
+                        for (const param of (member as ConstructorDeclaration).parameters) {
+                            if (isParameterPropertyDeclaration(param)) {
+                                addName(names, param, (param.name as Identifier).text, property);
+                            }
+                        }
+                        break;
+
+                    case SyntaxKind.GetAccessor:
+                        addName(names, member.name, memberName, getter);
+                        break;
+
+                    case SyntaxKind.SetAccessor:
+                        addName(names, member.name, memberName, setter);
+                        break;
+
+                    case SyntaxKind.PropertyDeclaration:
+                        addName(names, member.name, memberName, property);
+                        break;
+                }
+            }
+
+            function addName(names: Map<number>, location: Node, name: string, meaning: number) {
+                if (hasProperty(names, name)) {
+                    const prev = names[name];
+                    if (prev & meaning) {
+                        error(location, Diagnostics.Duplicate_identifier_0, name);
+                    }
+                    else {
+                        names[name] = prev | meaning;
+                    }
+                }
+                else {
+                    names[name] = meaning;
+                }
+            }
+        }
+
+        function checkObjectTypeForDuplicateDeclarations(node: TypeLiteralNode | InterfaceDeclaration) {
+            const names: Map<boolean> = {};
+            for (const member of node.members) {
+                if (member.kind == SyntaxKind.PropertySignature) {
+                    let memberName: string;
+                    switch (member.name.kind) {
+                        case SyntaxKind.StringLiteral:
+                        case SyntaxKind.NumericLiteral:
+                        case SyntaxKind.Identifier:
+                            memberName = (member.name as LiteralExpression | Identifier).text;
+                            break;
+                        default:
+                            continue;
+                    }
+
+                    if (hasProperty(names, memberName)) {
+                        error(member, Diagnostics.Duplicate_identifier_0, memberName);

```

Comment:
We report the duplicate declaration errors on all declarations, as ppl complained they could not find the other one.